### PR TITLE
[minor] Add authorize

### PR DESCRIPTION
**Pull request by ssunorz**
# Description

authorizerd.goにAuthorizeメソッドを追加

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Non-code changes (update documentation, pipeline, etc.)

### Flags

- [ ] breaks backward compatibility
- [x] requires a documentation update
- [ ] has untestable code

## Related issue/PR

**Delete if not applicable**
- Fixes #_issue_
- Closes #_PR_

---

## Checklist

- [x] Followed the guidelines in the CONTRIBUTING document
- [x] Added prefix `[major]`/`[minor]`/`[patch]`/`[skip]` in the PR title
- [x] Tested and linted the code
- [x] Commented the code
- [x] Made corresponding changes to the documentation
- [x] Confirmed no dropping in test coverage (by [Codecov](https://codecov.io/gh/yahoojapan/athenz-authorizer/pulls))
- [x] Passed all pipeline checking
- [x] Approved by >1 reviewer

## Checklist for maintainer
- [x] Use `Squash and merge`
- [x] Double-confirm the merge message has prefix `[major]`/`[minor]`/`[patch]`/`[skip]`
- [x] Delete the branch after merge

### CODE REVIEW

File: authorizerd.go [hunk 1]
```diff
@@ -331,77 +350,94 @@ func (a *authorizer) Start(ctx context.Context) <-chan error {
...
+	return err
+}
+
+// AuthorizeAccessToken verifies the access token on the specific (action, resource) pair and returns the result of verifying and verification error if unauthorized.
+func (a *authorizer) AuthorizeAccessToken(ctx context.Context, tok, act, res string, cert *x509.Certificate) (Principal, error) {
 	return a.verify(ctx, accessToken, tok, act, res, cert)
 }
 
-func (a *authorizer) verify(ctx context.Context, m mode, tok, act, res string, cert *x509.Certificate) error {
+func (a *authorizer) verify(ctx context.Context, m mode, tok, act, res string, cert *x509.Certificate) (Principal, error) {
```
Comment (WindzCUHK) [inline authorizerd.go:379]:
```suggestion
func (a *authorizer) authorize(ctx context.Context, m mode, tok, act, res string, cert *x509.Certificate) (Principal, error) {
```

File: role/token.go [hunk 2]
```diff
@@ -28,14 +28,16 @@ import (
...
+	Roles     []string // required
+	Principal string   // required
 	// Host       string
 	// Salt       string    // required
 	// TimeStamp  time.Time // required
-	ExpiryTime time.Time // required
-	KeyID      string    // required
+	IntTimeStamp  int64     // required
+	ExpiryTime    time.Time // required
+	IntExpiryTime int64     // requir_ed
```
Comment (WindzCUHK) [inline role/token.go:39]:
```suggestion
```
better to just have `ExpiryTime` and convert in the getter with `ExpiryTime.Unix()`?
same for `TimeStamp`

File: access/claim.go [hunk 3]
```diff
@@ -67,3 +67,33 @@ type OAuth2AccessTokenClaim struct {
...
+func (a *OAuth2AccessTokenClaim) GetRoles() []string {
+	return a.Scope
+}
+
+// GetDomain returns the the domain of the principal
+func (a *OAuth2AccessTokenClaim) GetDomain() string {
+	return a.BaseClaim.Audience
+}
+
+// GetIssueTime returns the the issued time of the principal
```
Comment (WindzCUHK) [inline access/claim.go:86]:
```suggestion
// GetIssueTime returns the the issued time of the principal (unix timestamp in second)
```

File: access/claim.go [hunk 4]
```diff
@@ -67,3 +67,33 @@ type OAuth2AccessTokenClaim struct {
...
+func (a *OAuth2AccessTokenClaim) GetIssueTime() int64 {
+	return a.BaseClaim.IssuedAt
+}
+
+// GetExpiryTime returns the the expiry time of the principal
```
Comment (WindzCUHK) [inline access/claim.go:91]:
```suggestion
// GetExpiryTime returns the the expiry time of the principal (unix timestamp in second)
```

File: authorizerd.go [hunk 5]
```diff
@@ -41,14 +41,33 @@ type Authorizerd interface {
...
+	AuthorizeRoleCert(ctx context.Context, peerCerts []*x509.Certificate, act, res string) (Principal, error)
 	GetPolicyCache(ctx context.Context) map[string]interface{}
 }
 
-type verifier func(r *http.Request, act, res string) error
+type verifier func(r *http.Request, act, res string) (Principal, error)
+
+// A Principal is an authenticated entity
+type Principal interface {
+	GetName() string
```
[the remark this answers is not in the document]
Comment (ssunorz) [reply authorizerd.go:59]: `Get`をなくすと、[token.go](https://github.com/yahoojapan/athenz-authorizer/blob/add-authorize/role/token.go#L31-L32)でfield名と重複してしまいます。

これを回避するため、field名を小文字にする（`domain`など）という提案がありましたが

publicだったものをprivateにするということだと思うので、使われている可能性は低いと思いますが、後方互換がなくなるように思います。

何か他によい提案はありますか？

[the remark this answers is not in the document]
Comment (WindzCUHK) [reply authorizerd.go:59]: For now, can only think out that having a private `type principal struct` that implement `Principal interface`.

And create the `principal` object in here.
https://github.com/yahoojapan/athenz-authorizer/blob/145b5796b8433091928dcd696ebd5fe89472c5d3/authorizerd.go#L404-L406

```go
principal = &principal{
    roles: rt.Roles
    ...
}
```

It is easier to refactor than changing the interface later on.

File: authorizerd.go [hunk 6]
```diff
@@ -41,14 +41,33 @@ type Authorizerd interface {
...
 	VerifyRoleToken(ctx context.Context, tok, act, res string) error
+	AuthorizeRoleToken(ctx context.Context, tok, act, res string) (Principal, error)
 	VerifyRoleJWT(ctx context.Context, tok, act, res string) error
 	VerifyRoleCert(ctx context.Context, peerCerts []*x509.Certificate, act, res string) error
+	AuthorizeRoleCert(ctx context.Context, peerCerts []*x509.Certificate, act, res string) (Principal, error)
 	GetPolicyCache(ctx context.Context) map[string]interface{}
 }
 
-type verifier func(r *http.Request, act, res string) error
+type verifier func(r *http.Request, act, res string) (Principal, error)
```
[the remark this answers is not in the document]
Comment (ssunorz) [reply authorizerd.go:55]: `authorizer`もconflictします
https://github.com/yahoojapan/athenz-authorizer/blob/145b5796b8433091928dcd696ebd5fe89472c5d3/authorizerd.go#L72

File: README.md [hunk 7]
```diff
@@ -62,6 +62,10 @@ go func() {
 if err := daemon.VerifyRoleToken(ctx, roleTok, act, res); err != nil {
     // token not authorized
 }
+
+// Verified results are returned
+principal, err := daemon.AuthorizeRoleToken(ctx, roleTok, act, res)
+name := principal.GetName()
```
[the remark this answers is not in the document]
Comment (ssunorz) [reply README.md]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/71/commits/8f2e7c54a9c8571f6eadc3e9cac6ab3d97f302ec

File: access/claim_test.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+*/
+package access
+
+import (
+	"github.com/dgrijalva/jwt-go"
+	"reflect"
+	"testing"
+)
+
+func TestOAuth2AccessTokenClaim_Principal(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ssunorz) [reply access/claim_test.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/71/commits/8f2e7c54a9c8571f6eadc3e9cac6ab3d97f302ec

File: authorizerd.go [hunk 9]
```diff
@@ -41,14 +41,33 @@ type Authorizerd interface {
...
+type Principal interface {
+	GetName() string
+	GetRoles() []string
+	GetDomain() string
+	GetIssueTime() int64
+	GetExpiryTime() int64
+}
+
+// A ClientID is an interface for a principal that has a ClientID
+type ClientID interface {
```
[the remark this answers is not in the document]
Comment (ssunorz) [reply authorizerd.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/71/commits/8f2e7c54a9c8571f6eadc3e9cac6ab3d97f302ec

File: authorizerd.go (hunk 1 above)
Comment (ssunorz) [reply authorizerd.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/71/commits/8f2e7c54a9c8571f6eadc3e9cac6ab3d97f302ec

File: role/token.go (hunk 2 above)
Comment (ssunorz) [reply role/token.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/71/commits/8f2e7c54a9c8571f6eadc3e9cac6ab3d97f302ec

File: access/claim.go (hunk 3 above)
Comment (ssunorz) [reply access/claim.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/71/commits/8f2e7c54a9c8571f6eadc3e9cac6ab3d97f302ec

File: access/claim.go (hunk 4 above)
Comment (ssunorz) [reply access/claim.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/71/commits/8f2e7c54a9c8571f6eadc3e9cac6ab3d97f302ec

File: authorizerd.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (WindzCUHK) [reply authorizerd.go:55]: how about `type authority struct {`?
One authority can have many authorizers.

[the remark this answers is not in the document]
Comment (ssunorz) [reply authorizerd.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/71/commits/fddfda9bdbb678cb9d184608ad8a7936293155b1

File: authorizerd.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ssunorz) [reply authorizerd.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/71/commits/fddfda9bdbb678cb9d184608ad8a7936293155b1

File: authorizerd.go [hunk 10]
```diff
@@ -410,8 +446,21 @@ func (a *authorizer) Verify(r *http.Request, act, res string) error {
 	return ErrInvalidCredentials
 }
 
+// Authorize returns the result of verifying and error of verification.
```
Comment (WindzCUHK) [inline authorizerd.go:449]:
```suggestion
// Authorize returns the principal or an error if unauthorized. Returns the principal with nil error if ANY authorizer succeeds (OR logic).
```

File: authorizerd.go [hunk 11]
```diff
@@ -447,7 +496,13 @@ func (a *authorizer) VerifyRoleCert(ctx context.Context, peerCerts []*x509.Certi
 	return errors.Wrap(err, "role certificates unauthorized")
 }
 
+// AuthorizeRoleCert verifies the role certificate for specific resource and return and verification error.
```
Comment (WindzCUHK) [inline authorizerd.go:499]:
```suggestion
// AuthorizeRoleCert verifies the role certificate for specific resource and returns the result of verifying or verification error if unauthorized. (unimplemented)
```

File: principal.go [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authorizerd
+
+// A Principal is an authenticated entity
```
Comment (WindzCUHK) [inline principal.go:19]:
```suggestion
// Principal is an authenticated entity
```

File: principal.go [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+type Principal interface {
+	Name() string
+	Roles() []string
+	Domain() string
+	IssueTime() int64
+	ExpiryTime() int64
+}
+
+// A OAuthAccessToken is an interface for a principal that has a OAuthAccessToken
```
Comment (WindzCUHK) [inline principal.go:28]:
```suggestion
// OAuthAccessToken is an interface for a principal that has a OAuthAccessToken
```

File: principal.go [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+	issueTime  int64
+	expiryTime int64
+}
+
+type oAuthAccessToken struct {
+	principal
+	clientID string
+}
+
+// Return the principal's name
```
Comment (WindzCUHK) [inline principal.go:46]:
```suggestion
// Name returns the principal's name
```
Just to be consistent with other comments, `<function name> <verb> <description>`
Please also help to change the comment of other getters.

File: principal.go [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+func (p *principal) IssueTime() int64 {
+	return p.issueTime
+}
+
+// Return the principal's expiryTime
+func (p *principal) ExpiryTime() int64 {
+	return p.expiryTime
+}
+
+// Return the principal's clientid
```
Comment (WindzCUHK) [inline principal.go:71]:
```suggestion
// ClientID returns the access token's client ID
```

File: role/token_test.go [hunk 16]
```diff
@@ -27,9 +27,12 @@ import (
...
 	type fields struct {
+		Principal     string
 		Domain        string
 		Roles         []string
+		IntTimeStamp  int64
 		ExpiryTime    time.Time
+		IntExpiryTime int64
 		KeyID         string
 		Signature     string
 		UnsignedToken string
```
Comment (WindzCUHK) [inline role/token_test.go:38]:
```suggestion
		Principal     string
		Domain        string
		Roles         []string
		TimeStamp  int64
		ExpiryTime    time.Time
		KeyID         string
		Signature     string
		UnsignedToken string
```
Please change this test to match the new fields

File: role/token.go [hunk 17]
```diff
@@ -62,18 +62,18 @@ func (r *Token) SetParams(key, value string) error {
...
-		// case "t":
-		// i, err := strconv.ParseInt(value, 10, 64)
-		// if err != nil {
-		// return err
-		// }
-		// r.TimeStamp = time.Unix(i, 0)
+	case "t":
+		i, err := strconv.ParseInt(value, 10, 64)
+		if err != nil {
+			return errors.Wrap(err, "invalid timestamp")
```
Comment (WindzCUHK) [inline role/token.go:74]: Please help to add a test case for this line.

File: authorizerd_test.go [hunk 18]
```diff
@@ -652,7 +652,7 @@ func Test_authorizer_Start(t *testing.T) {
 	}
 }
 
-func Test_authorizer_VerifyRoleToken(t *testing.T) {
+func Test_authorizer_VerifyRoleToken_AuthorizeRoleToken(t *testing.T) {
```
Comment (WindzCUHK) [inline authorizerd_test.go:655]:
```suggestion
func Test_authorizer_AuthorizeRoleToken(t *testing.T) {
```

File: authorizerd_test.go [hunk 19]
```diff
@@ -1409,15 +1459,15 @@ func Test_authorizer_GetPolicyCache(t *testing.T) {
...
 			if got := a.GetPolicyCache(tt.args.ctx); !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("authorizer.GetPolicyCache() = %v, want %v", got, tt.want)
+				t.Errorf("authority.GetPolicyCache() = %v, want %v", got, tt.want)
 			}
 		})
 	}
 }
 
-func Test_authorizer_Verify(t *testing.T) {
+func Test_authorizer_Verify_Authorize(t *testing.T) {
```
Comment (WindzCUHK) [inline authorizerd_test.go:1468]:
```suggestion
func Test_authorizer_Authorize(t *testing.T) {
```

File: authorizerd_test.go [hunk 20]
```diff
@@ -1490,16 +1540,20 @@ func Test_authorizer_Verify(t *testing.T) {
...
+			}
+			if _, err := a.Authorize(tt.args.r, tt.args.act, tt.args.res); (err != nil) != tt.wantErr {
+				t.Errorf("authority.Authorize() error = %v, wantErr %v", err, tt.wantErr)
 			}
 		})
 	}
 }
-func Test_authorizer_VerifyAccessToken(t *testing.T) {
+
+func Test_authorizer_VerifyAccessToken_AuthorizeAccessToken(t *testing.T) {
```
Comment (WindzCUHK) [inline authorizerd_test.go:1556]:
```suggestion
func Test_authorizer_AuthorizeAccessToken(t *testing.T) {
```

File: role/token_test.go (hunk 16 above)
Comment (ssunorz) [reply role/token_test.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/71/commits/95fccb9f7d6f4206dd1617486ac2abdbf7d36edc

File: role/token.go (hunk 17 above)
Comment (ssunorz) [reply role/token.go:74]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/71/commits/95fccb9f7d6f4206dd1617486ac2abdbf7d36edc

File: authorizerd_test.go (hunk 18 above)
Comment (ssunorz) [reply authorizerd_test.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/71/commits/95fccb9f7d6f4206dd1617486ac2abdbf7d36edc

File: authorizerd_test.go (hunk 19 above)
Comment (ssunorz) [reply authorizerd_test.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/71/commits/95fccb9f7d6f4206dd1617486ac2abdbf7d36edc

File: authorizerd_test.go (hunk 20 above)
Comment (ssunorz) [reply authorizerd_test.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/71/commits/95fccb9f7d6f4206dd1617486ac2abdbf7d36edc

**Pull request merged**